In `magicleap` build:

```
error[E0046]: not all trait items implemented, missing: `on_allow_navigation`
   --> ports/libmlservo/src/lib.rs:338:1
    |
338 | impl HostTrait for HostCallbacks {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `on_allow_navigation` in implementation
    |
    = note: `on_allow_navigation` from trait: `fn(&Self, std::string::String) -> bool`

error: aborting due to previous error
```

--- a/ports/libmlservo/src/lib.rs
+++ b/ports/libmlservo/src/lib.rs
@@ -350,6 +350,9 @@ impl HostTrait for HostCallbacks {
     fn on_load_started(&self) {}
     fn on_load_ended(&self) {}
     fn on_title_changed(&self, _title: String) {}
+    fn on_allow_navigation(&self, _url: String) -> bool {
+        true
+    }
     fn on_url_changed(&self, url: String) {
         if let Ok(cstr) = CString::new(url.as_str()) {
             if let Some(url_update) = self.url_update.0 {

--- a/ports/libsimpleservo/api/src/lib.rs
+++ b/ports/libsimpleservo/api/src/lib.rs
@@ -97,6 +97,8 @@ pub trait HostTrait {
     fn on_load_ended(&self);
     /// Page title has changed.
     fn on_title_changed(&self, title: String);
+    /// Allow Navigation.
+    fn on_allow_navigation(&self, url: String) -> bool;
     /// Page URL has changed.
     fn on_url_changed(&self, url: String);
     /// Back/forward state has changed.
@@ -474,9 +476,13 @@ impl ServoGlue {
                     let title = format!("{} - Servo", title);
                     self.callbacks.host_callbacks.on_title_changed(title);
                 },
-                EmbedderMsg::AllowNavigationRequest(pipeline_id, _url) => {
+                EmbedderMsg::AllowNavigationRequest(pipeline_id, url) => {
                     if let Some(_browser_id) = browser_id {
-                        let window_event = WindowEvent::AllowNavigationResponse(pipeline_id, true);
+                        let data: bool = self
+                            .callbacks
+                            .host_callbacks
+                            .on_allow_navigation(url.to_string());
+                        let window_event = WindowEvent::AllowNavigationResponse(pipeline_id, data);
                         let _ = self.process_event(window_event);
                     }
                 },

--- a/ports/libsimpleservo/capi/src/lib.rs
+++ b/ports/libsimpleservo/capi/src/lib.rs
@@ -37,6 +37,7 @@ pub struct CHostCallbacks {
     pub on_load_started: extern "C" fn(),
     pub on_load_ended: extern "C" fn(),
     pub on_title_changed: extern "C" fn(title: *const c_char),
+    pub on_allow_navigation: extern "C" fn(url: *const c_char) -> bool,
     pub on_url_changed: extern "C" fn(url: *const c_char),
     pub on_history_changed: extern "C" fn(can_go_back: bool, can_go_forward: bool),
     pub on_animating_changed: extern "C" fn(animating: bool),
@@ -320,6 +321,14 @@ impl HostTrait for HostCallbacks {
         let title_ptr = title.as_ptr();
         mem::forget(title);
         (self.0.on_title_changed)(title_ptr);
+    }
+
+    fn on_allow_navigation(&self, url: String) -> bool {
+        debug!("on_allow_navigation");
+        let url = CString::new(url).expect("Can't create string");
+        let url_ptr = url.as_ptr();
+        mem::forget(url);
+        (self.0.on_allow_navigation)(url_ptr)
     }
 
     fn on_url_changed(&self, url: String) {

--- a/ports/libsimpleservo/jniapi/src/lib.rs
+++ b/ports/libsimpleservo/jniapi/src/lib.rs
@@ -411,6 +411,26 @@ impl HostTrait for HostCallbacks {
         .unwrap();
     }
 
+    fn on_allow_navigation(&self, url: String) -> bool {
+        debug!("on_allow_navigation");
+        let env = self.jvm.get_env().unwrap();
+        let s = match new_string(&env, &url) {
+            Ok(s) => s,
+            Err(_) => return false,
+        };
+        let s = JValue::from(JObject::from(s));
+        let allow = env.call_method(
+            self.callbacks.as_obj(),
+            "onAllowNavigation",
+            "(Ljava/lang/String;)Z",
+            &[s],
+        );
+        match allow {
+            Ok(allow) => return allow.z().unwrap(),
+            Err(_) => return true,
+        }
+    }
+
     fn on_url_changed(&self, url: String) {
         debug!("on_url_changed");
         let env = self.jvm.get_env().unwrap();

--- a/support/android/apk/servoapp/src/main/java/org/mozilla/servo/MainActivity.java
+++ b/support/android/apk/servoapp/src/main/java/org/mozilla/servo/MainActivity.java
@@ -20,6 +20,7 @@ import android.widget.Button;
 import android.widget.EditText;
 import android.widget.ProgressBar;
 import android.widget.TextView;
+import android.util.Log;
 
 import org.mozilla.servoview.ServoView;
 import org.mozilla.servoview.Servo;
@@ -160,6 +161,19 @@ public class MainActivity extends Activity implements Servo.Client {
         mBackButton.setEnabled(canGoBack);
         mFwdButton.setEnabled(canGoForward);
         mCanGoBack = canGoBack;
+    }
+
+    @Override
+    public boolean onAllowNavigation(String url) {
+        if (url.startsWith("market://")) {
+            try {
+                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(url)));
+                return false;
+            } catch (Exception e) {
+                Log.e("onAllowNavigation", e.toString());
+            }
+        }
+        return true;
     }
 
     public void onRedrawing(boolean redrawing) {

--- a/support/android/apk/servoview/src/main/java/org/mozilla/servoview/Servo.java
+++ b/support/android/apk/servoview/src/main/java/org/mozilla/servoview/Servo.java
@@ -169,6 +169,8 @@ public class Servo {
     }
 
     public interface Client {
+        boolean onAllowNavigation(String url);
+
         void onLoadStarted();
 
         void onLoadEnded();
@@ -232,6 +234,10 @@ public class Servo {
 
         public void onAnimatingChanged(boolean animating) {
             mRunCallback.inGLThread(() -> mGfxCb.animationStateChanged(animating));
+        }
+
+        public boolean onAllowNavigation(String url) {
+            return mClient.onAllowNavigation(url);
         }
 
         public void onLoadStarted() {

